src/zarr/store/remote.py

```diff
@@ -1,103 +1,190 @@
+                )
+            self.path = url.path
+            self._fs = url._fs
         else:
-            assert (
-                len(storage_options) == 0
-            ), "If constructed with a UPath object, no additional storage_options are allowed."
-            self.root = url.rstrip("/")
-
+            raise ValueError("URL not understood, %s", url)
+        self.exceptions = allowed_exceptions
         # test instantiate file system
-        fs, _ = fsspec.core.url_to_fs(str(self.root), asynchronous=True, **self.root._kwargs)
-        assert fs.__class__.async_impl, "FileSystem needs to support async operations."
+        if not self._fs.async_impl:
+            raise TypeError("FileSystem needs to support async operations")
 
     def __str__(self) -> str:
-        return str(self.root)
+        return f"Remote fsspec store: {type(self._fs).__name__} , {self.path}"
 
     def __repr__(self) -> str:
-        return f"RemoteStore({str(self)!r})"
-
-    def _make_fs(self) -> tuple[AsyncFileSystem, str]:
-        import fsspec
-
-        storage_options = self.root._kwargs.copy()
-        storage_options.pop("_url", None)
-        fs, root = fsspec.core.url_to_fs(str(self.root), asynchronous=True, **self.root._kwargs)
-        assert fs.__class__.async_impl, "FileSystem needs to support async operations."
-        return fs, root
+        return f"<RemoteStore({type(self._fs).__name__} , {self.path})>"
 
     async def get(
         self,
         key: str,
-        prototype: BufferPrototype,
+        prototype: BufferPrototype = default_buffer_prototype,
         byte_range: tuple[int | None, int | None] | None = None,
     ) -> Buffer | None:
-        assert isinstance(key, str)
-        fs, root = self._make_fs()
-        path = _dereference_path(root, key)
+        path = _dereference_path(self.path, key)
 
         try:
-            value: Buffer | None = await (
-                fs._cat_file(path, start=byte_range[0], end=byte_range[1])
-                if byte_range
-                else fs._cat_file(path)
+            if byte_range:
+                # fsspec uses start/end, not start/length
+                start, length = byte_range
+                if start is not None and length is not None:
+                    end = start + length
+                elif length is not None:
+                    end = length
+                else:
+                    end = None
+            value: Buffer = prototype.buffer.from_bytes(
+                await (
+                    self._fs._cat_file(path, start=byte_range[0], end=end)
+                    if byte_range
+                    else self._fs._cat_file(path)
+                )
             )
-        except (FileNotFoundError, IsADirectoryError, NotADirectoryError):
-            return None
+            return value
 
-        return value
+        except self.exceptions:
+            return None
+        except OSError as e:
+            if "not satisfiable" in str(e):
```

Comment:
flagging this as s3 abstraction leakage that we might want to address later on by making an s3-specific storage class